hypertrace-graphql-labels-schema/src/main/java/org/hypertrace/graphql/label/joiner/DefaultLabelJoinerBuilder.java

```diff
@@ -0,0 +1,97 @@
+package org.hypertrace.graphql.label.joiner;
+
+import static java.util.function.Function.identity;
+
+import io.reactivex.rxjava3.core.Observable;
+import io.reactivex.rxjava3.core.Single;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import javax.inject.Inject;
+import lombok.AllArgsConstructor;
+import lombok.experimental.Accessors;
+import lombok.extern.slf4j.Slf4j;
+import org.hypertrace.core.graphql.context.GraphQlRequestContext;
+import org.hypertrace.graphql.label.dao.LabelDao;
+import org.hypertrace.graphql.label.schema.LabelResultSet;
+import org.hypertrace.label.config.service.v1.GetLabelsRequest;
+import org.hypertrace.label.config.service.v1.GetLabelsResponse;
+import org.hypertrace.label.config.service.v1.Label;
+
+@Slf4j
+class DefaultLabelJoinerBuilder implements LabelJoinerBuilder {
+
+  private final LabelDao labelDao;
+
+  @Inject
+  DefaultLabelJoinerBuilder(LabelDao labelDao) {
+    this.labelDao = labelDao;
+  }
+
+  @Override
+  public Single<LabelJoiner> build(GraphQlRequestContext context) {
+    return Single.just(new DefaultLabelJoiner(context));
+  }
+
+  @AllArgsConstructor
+  private class DefaultLabelJoiner implements LabelJoiner {
+    private final GraphQlRequestContext context;
+
+    @Override
+    public <T> Single<Map<String, LabelResultSet>> joinLabels(
+        Collection<T> joinSources, IdGetter<T> idGetter, LabelsGetter<T> labelsGetter) {
+      return labelDao
+          .getLabelsResponse(context, GetLabelsRequest.getDefaultInstance())
+          .map(this::getLabelsMap)
+          .flatMap(labelsMap -> getResultSetMap(labelsMap, joinSources, idGetter, labelsGetter));
+    }
+
+    private Map<String, Label> getLabelsMap(GetLabelsResponse labelsResponse) {
+      return labelsResponse.getLabelsList().stream()
+          .collect(Collectors.toUnmodifiableMap(Label::getId, identity()));
+    }
+
+    private <T> Single<Map<String, LabelResultSet>> getResultSetMap(
+        Map<String, Label> labelsMap,
+        Collection<T> joinSources,
+        IdGetter<T> idGetter,
+        LabelsGetter<T> labelsGetter) {
+      return Observable.fromIterable(joinSources)
+          .toMap(
+              source -> idGetter.getId(source),
+              source -> buildLabelResultSet(labelsGetter.getLabels(source), labelsMap));
+    }
+
+    private LabelResultSet buildLabelResultSet(
+        List<String> labelIds, Map<String, Label> labelsMap) {
+      if (labelIds.isEmpty() || labelsMap.isEmpty()) {
+        return ConvertedLabelResultSet.EMPTY_LABEL_RESULT_SET;
+      }
+      List<org.hypertrace.graphql.label.schema.Label> convertedLabels =
+          labelsMap.entrySet().stream()
```

Comment:
same comment from last round here - anchor on the requested IDs rather than the returned IDs so we can catch if any requested label is missing.

Reply:
I have removed map and using a simple list. If this is needed, then I can bring back the map and do it.

Do we want to log missing label Ids?

Reply:
yes (continued this thread in the new location)